cluster-autoscaler/core/scale_down.go

```diff
@@ -384,6 +387,8 @@ func NewScaleDown(context *context.AutoscalingContext, clusterStateRegistry *clu
 
 // CleanUp cleans up the internal ScaleDown state.
 func (sd *ScaleDown) CleanUp(timestamp time.Time) {
+	// Use default ScaleDownUnneededTime as in this context the value
+	// doesn't apply to any specific NodeGroup.
 	sd.usageTracker.CleanUp(timestamp.Add(-sd.context.ScaleDownUnneededTime))
```

Comment:
You made me read usage tracker again (just when I *almost* managed to forget the horrors that lie within...). :(

I don't see any reason why it should be minimum. When we actually scale-down node we remove it explicitly from tracker, we don't need CleanUp to do it for us. If anything, setting this value to be too small will wipe usage information from tracker effectively preventing it from doing anything.

That being said - it doesn't make any sense to me that this is tied to ScaleDownUnneededTime at all. After going through the code I think we record all usage in tracker each time we run UpdateUnneededNodes (usageTracker is updated in FindNodesToRemove whether using hint or not and we call FindNodesToRemove for both currentCandidates and when seeking additional candidates). The threshold for CleanUp should really be set in a way that we remove all records that were not created in the last UpdateUnneededNodes call. This is tricky as it's related to loop time (the actual loop time and not --scan-interval) adjusted for whether we triggered scale-up or not (we don't call UpdateUnneededNodes in a loop when we trigger scale-up). 
I don't think any threshold can achieve this. I think a better way to do this would be to use some sort of logical clock that would be incremented every time we call UpdateUnneededNodes. However, this is a bigger change that isn't completely unrelated to this PR. I'm leaning towards leaving it as-is:
* It maintains the current behavior.
* There is no good value to use. Based on reading code and educated guesses (I haven't actually run any tests):
  1. Small value (1m) in a large cluster seems to have realistic risk of skipping tracker logic completely by wiping its records clean every loop (loop time >1m seems pretty realistic to me). This introduces a small risk of overzealous scale-up with rapid scale-down config.
  2. Large value means that when we actually scale-down we purge unnededNodes list based on tracker data from old simulations that are completely irrelevant. This can slow down the scale-down.
  3. I'm against using minimum because of (i) seems like the wrong direction to fail in cluster with mixed configs. I think a case could be made for both using default and using max across node groups. With (i) seeming much less likely to actually trigger than (ii) I'm leaning towards using the default.

Reply:
Thanks for the detailed explanation, leaving as-is SGTM!

Reply:
It just occurred to me that I may have been missing the obvious here. Why can't we just use 0? Based on quickly looking at the code:
1. We built all relevant information in tracker in UpdateUnneededNodes.
2. We use this information to reset timer for some candidates when performing scale-down.
3. TryToScaleDown is called after running UpdateUnneededNodes. We never call TryToScaleDown without calling UpdateUnneededNodes first.

If those 3 points above are correct - why exactly do we need to keep anything in tracker between loops? Why not just fully reset it at the start of scale-down? Am I missing something? Looking for a sanity check.

Reply:
When I looked for the scalability issue in scale-down I also noticed that. Basically IMO the timestamps stored by UsageTracker have *no real effect* on the mechanism of usageTracker - all the relevant timestamps are stored in ScaleDown.unremovableNodes either way. The usageTracker is also rebuild in each scale-down loop, so I think that your intuition is correct.